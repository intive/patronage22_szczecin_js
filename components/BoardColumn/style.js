import styled from 'styled-components'
import Button from '../Button/Button'
import Icon from '../Icon/Icon'

export const Column = styled.div`
  width: 328px;
`

export const ColumnCard = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  margin: 1rem 0;
  padding: ${props => props.card ? '1rem' : '1.25rem 1rem 0.75rem 1rem'};
  border-radius: 8px;
  background-color: #ffffff;
`

export const ColumnCardTitle = styled.h4`
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.5rem;
  margin-bottom: 1.25rem;
  color: #7E7E7E;
`

export const StyledButton = styled(Button)`
  padding: 0.5rem;
  width: ${props => props.outline ? '100%' : 'auto'};
`

export const ColumnCardText = styled.div`
  width: 100%;
  padding: 0.75rem 1rem;
  border: 1px solid #E1E1E1;
  border-radius: 8px;
  background: #FFFFFF;
  overflow-wrap: break-word;
`

export const ColumnAddCardWrapper = styled.div`
  width: 100%;
  display: flex;
  justify-content: ${props => props.right ? 'flex-end' : 'space-between'};
  align-items: center;
  gap: 12px;
  margin-top: ${props => props.right ? '0' : '1rem'}
`

export const ColumnCardName = styled.span`
  font-size: 0.75rem;
  font-weight: 400;
  line-height: 1rem;
  color: #7E7E7E;
`

export const StyledIcon = styled(Icon)`
  line-height: 1rem;
  color: #7E7E7E;
  cursor: pointer;
  &:hover{
    color: #515151;
  }
`
