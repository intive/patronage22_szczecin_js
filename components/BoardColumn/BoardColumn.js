import { StyledButton, Column, ColumnCardTitle, ColumnAddCardWrapper, ColumnCard, ColumnCardText, ColumnCardName, StyledIcon } from './style'
import { useState, useReducer, useRef, useEffect } from 'react'
import { useTranslation } from 'next-i18next'
import { ulid } from 'ulid'
import TextArea from '../TextArea/TextArea'

const BoardColumn = (props) => {
  const { t } = useTranslation('common')

  function init (initialText) {
    return { cardTextValue: initialText }
  }

  function reducer (state, action) {
    switch (action.type) {
      case 'change':
        return { ...state, cardTextValue: action.payload }
      case 'save':
        return { cardTextValue: '' }
      case 'reset':
        return init(action.payload)
      default:
        throw new Error()
    }
  }

  const [isInCreateCardMode, setIsInCreateCardMode] = useState(false)
  const [cards, setCards] = useState(props.cards || [])

  const [state, dispatch] = useReducer(reducer, init)
  const { cardTextValue } = state

  const textAreaRef = useRef(null)

  useEffect(() => {
    if (isInCreateCardMode) {
      textAreaRef.current.focus()
    }
  }, [isInCreateCardMode])

  const handleOnChange = (event) => {
    dispatch({ type: 'change', payload: event.target.value })
  }

  const handleSaveCard = (event) => {
    event.preventDefault()
    dispatch({ type: 'save' })
    addCard(event)
    setIsInCreateCardMode(false)
  }

  const handleCancelCard = (event) => {
    event.preventDefault()
    dispatch({ type: 'reset' })
    setIsInCreateCardMode(false)
  }

  const switchToCreateCardMode = () => {
    setIsInCreateCardMode(true)
  }

  const addCard = () => {
    setCards([
      ...cards,
      {
        id: ulid(),
        text: cardTextValue,
        author: 'Isaak Newton'
      }
    ])
  }

  return (
    <Column data-testid='board-column'>
      <ColumnCard>
        <ColumnCardTitle>{props.name}</ColumnCardTitle>
        {isInCreateCardMode
          ? (
            <>
              <TextArea value={cardTextValue} name='cardTextValue' onChange={handleOnChange} ref={textAreaRef} maxLength='300' />
              <ColumnAddCardWrapper right>
                <StyledButton text onClick={handleCancelCard}>{t('buttons.cancel')}</StyledButton>
                <StyledButton text disabled={!cardTextValue || cardTextValue.trim().length < 5} onClick={handleSaveCard}>{t('buttons.save')}</StyledButton>
              </ColumnAddCardWrapper>
            </>
            )
          : <StyledButton text outline onClick={switchToCreateCardMode}>{t('boardColumn.addCard')}</StyledButton>}
      </ColumnCard>
      {cards
        ? (
          <div>
            {cards.map((card) => (
              <ColumnCard key={card.id} card>
                <ColumnCardText>{card.text}</ColumnCardText>
                <ColumnAddCardWrapper>
                  <ColumnCardName>{card.author}</ColumnCardName>
                  <StyledIcon name='favorite_border' />
                </ColumnAddCardWrapper>
              </ColumnCard>
            ))}
          </div>
          )
        : null}
    </Column>
  )
}

export default BoardColumn
