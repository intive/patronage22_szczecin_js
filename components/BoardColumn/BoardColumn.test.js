/* globals describe, expect, it */

import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import BoardColumn from './BoardColumn'

describe('BoardColumn', () => {
  it('should render BoardColumn component', () => {
    const { container } = render(<BoardColumn />)

    expect(container).toMatchSnapshot()
  })

  it('should hide add card button and show cancel button after click on add card button', () => {
    render(<BoardColumn />)
    const button = screen.getByText('boardColumn.addCard')
    userEvent.click(button)

    expect(screen.queryByText(/buttons.cancel/i)).toBeVisible()
    expect(button).not.toBeVisible()
  })

  it('should hide add card button and show save button after click on add card button', () => {
    render(<BoardColumn />)
    const button = screen.getByText('boardColumn.addCard')
    userEvent.click(button)

    expect(screen.queryByText(/buttons.save/i)).toBeVisible()
    expect(button).not.toBeInTheDocument()
  })

  it('should hide save button and show add card button after click on save button', () => {
    render(<BoardColumn />)
    userEvent.click(screen.getByText('boardColumn.addCard'))
    userEvent.click(screen.getByText('buttons.save'))

    expect(screen.queryByText(/boardColumn.save/i)).not.toBeInTheDocument()
    expect(screen.getByText('boardColumn.addCard')).toBeVisible()
  })

  it('should hide cancel button and show add card button after click on cancel button', () => {
    render(<BoardColumn />)
    userEvent.click(screen.getByText('boardColumn.addCard'))
    userEvent.click(screen.getByText('buttons.cancel'))

    expect(screen.queryByText(/boardColumn.cancel/i)).not.toBeInTheDocument()
    expect(screen.getByText('boardColumn.addCard')).toBeVisible()
  })
})
