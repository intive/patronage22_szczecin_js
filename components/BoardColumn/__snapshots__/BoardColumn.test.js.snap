// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardColumn should render BoardColumn component 1`] = `
<div>
  <div
    class="sc-dkPtRN gUlzXt"
    data-testid="board-column"
  >
    <div
      class="sc-hKwDye iYNrnp"
    >
      <h4
        class="sc-eCImPb gGbVhf"
      >
        Kolumna 1
      </h4>
      <button
        class="sc-bdvvtL boJPqq sc-jRQBWg ishcL"
      >
        boardColumn.addCard
      </button>
    </div>
  </div>
</div>
`;
